autogen/oai/gemini.py

```diff
@@ -0,0 +1,225 @@
+"""Create a OpenAI-compatible client for Gemini features.
+
+Resources:
+- https://ai.google.dev/docs
+- https://cloud.google.com/vertex-ai/docs/generative-ai/migrate-from-azure
+- https://blog.google/technology/ai/google-gemini-pro-imagen-duet-ai-update/
+- https://ai.google.dev/api/python/google/generativeai/ChatSession
+"""
+
+from __future__ import annotations
+
+import os
+import pdb
+import random
+import time
+from typing import Any, Dict, List, Mapping, Union
+
+import google.generativeai as genai
+import httpx
+from google.ai.generativelanguage import Content, Part
+from google.generativeai import ChatSession
+from openai import OpenAI, _exceptions, resources
+from openai._qs import Querystring
+from openai._types import NOT_GIVEN, NotGiven, Omit, ProxiesTypes, RequestOptions, Timeout, Transport
+from openai.types.chat import ChatCompletion
+from openai.types.chat.chat_completion import ChatCompletionMessage, Choice
+from openai.types.completion_usage import CompletionUsage
+from proto.marshal.collections.repeated import RepeatedComposite
+from pydash import max_
+from typing_extensions import Self, override
+
+from autogen.token_count_utils import count_token
+
+
+class GeminiClient:
+    """
+    _summary_
+
+    _extended_summary_
+
+    TODO: this Gemini implementation does not support the following features yet:
+    - function_call (OpenAI)
+    - tool_calls (OpenAI)
+    - safety_setting (Gemini)
+    - Multi-turn chat for vision model (Gemini)
+    - multiple responses at the same time (Gemini)
+    """
+
+    def __init__(self, **kwargs):
+        self.api_key = kwargs.get("api_key", None)
+        if self.api_key is None:
+            self.api_key = os.getenv("GOOGLE_API_KEY")
+
+        assert self.api_key is not None, (
+            "Please provide api_key in OAI_CONFIG_LIST " "or set the GOOGLE_API_KEY env variable."
+        )
+
+    def call(self, params: Dict) -> ChatCompletion:
+        model_name = params.get("model", "gemini-pro")
+        params.get("api_type", "google")  # not used
+        messages = params.get("messages", [])
+        stream = params.get("stream", False)
+        n_response = params.get("n", 1)
+        params.get("temperature", 0.5)
+        params.get("top_p", 1.0)
+        params.get("max_tokens", 1024)
```

Comment:
Couldn't these be used by Gemini?

Reply:
They can! I just need more work to include them (that's why this PR is still a draft). 

Hold tight, and I will include them~!